packages/toc/src/toc_item.tsx

```diff
@@ -2,7 +2,120 @@
 // Distributed under the terms of the Modified BSD License.
 
 import * as React from 'react';
-import { IHeading } from './utils/headings';
+import * as ts from 'typescript';
+import { CommandRegistry } from '@lumino/commands';
+import { IHeading, INotebookHeading } from './utils/headings';
+import { Menu } from '@lumino/widgets';
+import { CodeCell } from '@jupyterlab/cells';
+import { NotebookPanel } from '@jupyterlab/notebook';
+import { CommandIDs } from './toc_tree';
+
+/**
+ * Adds commands to command registry.
+ *
+ * @param commands - command registry
+ * @param toc - List of headings
+ * @param heading - heading at current position
+ * @returns
+ */
+const addCommands = (
+  commands: CommandRegistry,
+  toc: IHeading[],
+  heading: IHeading
+) => {
+  commands.addCommand(CommandIDs.runCells, {
+    execute: args => {
+      const runCells = args['runCells'] as string;
+      const func = ts.transpile(runCells);
+      const runnable = eval(func);
+      let code: INotebookHeading[];
+      code = [];
+      return runnable(toc, heading, code);
+    },
+    label: 'Run Cell(s)'
+  });
+};
```

Comment:
A very rough draft, probably has linting issues:

```suggestion
const addCommands = (commands: CommandRegistry) => {
  commands.addCommand(CommandIDs.runCells, {
    execute: args => {
      const toc = args['toc'] as IHeading[];
      const code = args['code'] as INotebookHeading[];
      const heading = args['heading'] as IHeading;
      return NestedCodeCells(toc, heading, code);
    },
    label: 'Run Cell(s)'
  });
};
```

And then call `addCommands` function only once. It might not be as easy (because there might be serialization issues) but dealing with such issues (by modifying what arguments you end up passing in ) seems less problematic than re-defining the command at each render call.

Passing function and running eval to me (which should not be needed anyways).